README.md
# gitlab-workhorse

Gitlab-workhorse is a smart reverse proxy for GitLab. It handles
"large" HTTP requests such as file downloads, file uploads, Git
push/pull and Git archive downloads.

## Quick facts (how does Workhorse work)

-   Workhorse can handle some requests without involving Rails at all:
    for example, Javascript files and CSS files are served straight
    from disk.
-   Workhorse can modify responses sent by Rails: for example if you use
    `send_file` in Rails then gitlab-workhorse will open the file on
    disk and send its contents as the response body to the client.
-   Workhorse can take over requests after asking permission from Rails.
    Example: handling `git clone`.
-   Workhorse can modify requests before passing them to Rails. Example:
    when handling a Git LFS upload Workhorse first asks permission from
    Rails, then it stores the request body in a tempfile, then it sends
    a modified request containing the tempfile path to Rails.
-   Workhorse can manage long-lived WebSocket connections for Rails.
    Example: handling the terminal websocket for environments.
-   Workhorse does not connect to Postgres, only to Rails and (optionally) Redis.
-   We assume that all requests that reach Workhorse pass through an
    upstream proxy such as NGINX or Apache first.
-   Workhorse does not accept HTTPS connections.
-   Workhorse does not clean up idle client connections.
-   We assume that all requests to Rails pass through Workhorse.

For more information see ['A brief history of
gitlab-workhorse'][brief-history-blog].

## Usage

```
  gitlab-workhorse [OPTIONS]

Options:
  -apiCiLongPollingDuration duration
      Long polling duration for job requesting for runners (default 50s - enabled) (default 50ns)
  -apiLimit uint
      Number of API requests allowed at single time
  -apiQueueDuration duration
      Maximum queueing duration of requests (default 30s)
  -apiQueueLimit uint
      Number of API requests allowed to be queued
  -authBackend string
      Authentication/authorization backend (default "http://localhost:8080")
  -authSocket string
      Optional: Unix domain socket to dial authBackend at
  -config string
      TOML file to load config from
  -developmentMode
      Allow the assets to be served from Rails app
  -documentRoot string
      Path to static files content (default "public")
  -listenAddr string
      Listen address for HTTP server (default "localhost:8181")
  -listenNetwork string
      Listen 'network' (tcp, tcp4, tcp6, unix) (default "tcp")
  -listenUmask int
      Umask for Unix socket
  -logFile string
      Log file location
  -logFormat string
      Log format to use defaults to text (text, json, structured, none) (default "text")
  -pprofListenAddr string
      pprof listening address, e.g. 'localhost:6060'
  -prometheusListenAddr string
      Prometheus listening address, e.g. 'localhost:9229'
  -proxyHeadersTimeout duration
      How long to wait for response headers when proxying the request (default 5m0s)
  -secretPath string
      File with secret key to authenticate with authBackend (default "./.gitlab_workhorse_secret")
  -version
      Print version and exit
```

The 'auth backend' refers to the GitLab Rails application. The name is
a holdover from when gitlab-workhorse only handled Git push/pull over
HTTP.

Gitlab-workhorse can listen on either a TCP or a Unix domain socket. It
can also open a second listening TCP listening socket with the Go
[net/http/pprof profiler server](http://golang.org/pkg/net/http/pprof/).

Gitlab-workhorse can listen on redis events (currently only builds/register
for runners). This requires you to pass a valid TOML config file via
`-config` flag.
For regular setups it only requires the following (replacing the string
with the actual socket)

### Redis

Gitlab-workhorse integrates with Redis to do long polling for CI build
requests. This is configured via two things:

-   Redis settings in the TOML config file
-   The `-apiCiLongPollingDuration` command line flag to control polling
    behavior for CI build requests

It is OK to enable Redis in the config file but to leave CI polling
disabled; this just results in an idle Redis pubsub connection. The
opposite is not possible: CI long polling requires a correct Redis
configuration.

Below we discuss the options for the `[redis]` section in the config
file.

```
[redis]
URL = "unix:///var/run/gitlab/redis.sock"
Password = "my_awesome_password"
Sentinel = [ "tcp://sentinel1:23456", "tcp://sentinel2:23456" ]
SentinelMaster = "mymaster"
```

- `URL` takes a string in the format `unix://path/to/redis.sock` or
`tcp://host:port`.
- `Password` is only required if your redis instance is password-protected
- `Sentinel` is used if you are using Sentinel.
  *NOTE* that if both `Sentinel` and `URL` are given, only `Sentinel` will be used

Optional fields are as follows:
```
[redis]
DB = 0
ReadTimeout = "1s"
KeepAlivePeriod = "5m"
MaxIdle = 1
MaxActive = 1
```

- `DB` is the Database to connect to. Defaults to `0`
- `ReadTimeout` is how long a redis read-command can take. Defaults to `1s`
- `KeepAlivePeriod` is how long the redis connection is to be kept alive without anything flowing through it. Defaults to `5m`
- `MaxIdle` is how many idle connections can be in the redis-pool at once. Defaults to 1
- `MaxActive` is how many connections the pool can keep. Defaults to 1

### Health-checks

The Workhorse exposes it's own internal healthchecks on the `-prometheusListenAddr`:

- `/liveness` endpoint that always responds with OK when running
- `/readiness` endpoint that responds with OK when all downstream probes are successful

In order to add downstream probes for `/readiness` endpoint
add a `[readiness]` in the config file.

```toml
[readiness]
ProbesURL = [
  "http://localhost:8083/readiness"
]
```

### Relative URL support

If you are mounting GitLab at a relative URL, e.g.
`example.com/gitlab`, then you should also use this relative URL in
the `authBackend` setting:

```
gitlab-workhorse -authBackend http://localhost:8080/gitlab
```

## Installation

To install gitlab-workhorse you need [Go 1.8 or
newer](https://golang.org/dl) and [GNU
Make](https://www.gnu.org/software/make/).

To install into `/usr/local/bin` run `make install`.

```
make install
```

To install into `/foo/bin` set the PREFIX variable.

```
make install PREFIX=/foo
```

On some operating systems, such as FreeBSD, you may have to use
`gmake` instead of `make`.

## Dependencies

### Exiftool

Workhorse uses [exiftool](https://www.sno.phy.queensu.ca/~phil/exiftool/) for
removing EXIF data (which may contain sensitive information) from uploaded
images. If you installed GitLab:

-   Using the Omnibus package, you're all set.
    *NOTE* that if you are using CentOS Minimal, you may need to install `perl`
    package: `yum install perl`
-   From source, make sure `exiftool` is installed:

    ```sh
    # Debian/Ubuntu
    sudo apt-get install libimage-exiftool-perl

    # RHEL/CentOS
    sudo yum install perl-Image-ExifTool
    ```

## Error tracking

GitLab-Workhorse supports remote error tracking with
[Sentry](https://sentry.io). To enable this feature set the
GITLAB_WORKHORSE_SENTRY_DSN environment variable.
You can also set the GITLAB_WORKHORSE_SENTRY_ENVIRONMENT environment variable to
use the Sentry environment functionality to separate staging, production and
development.

Omnibus (`/etc/gitlab/gitlab.rb`):

```
gitlab_workhorse['env'] = {
    'GITLAB_WORKHORSE_SENTRY_DSN' => 'https://foobar'
    'GITLAB_WORKHORSE_SENTRY_ENVIRONMENT' => 'production'
}
```

Source installations (`/etc/default/gitlab`):

```
export GITLAB_WORKHORSE_SENTRY_DSN='https://foobar'
export GITLAB_WORKHORSE_SENTRY_ENVIRONMENT='production'
```

## Tests

Run the tests with:

```
make clean test
```

### Coverage / what to test

Each feature in gitlab-workhorse should have an integration test that
verifies that the feature 'kicks in' on the right requests and leaves
other requests unaffected. It is better to also have package-level tests
for specific behavior but the high-level integration tests should have
the first priority during development.

It is OK if a feature is only covered by integration tests.

## Distributed Tracing

Workhorse supports distributed tracing through [LabKit](https://gitlab.com/gitlab-org/labkit/) using [OpenTracing APIs](https://opentracing.io).

By default, no tracing implementation is linked into the binary, but different OpenTracing providers can be linked in using [build tags](https://golang.org/pkg/go/build/#hdr-Build_Constraints)/[build constraints](https://golang.org/pkg/go/build/#hdr-Build_Constraints). This can be done by setting the `BUILD_TAGS` make variable.

For more details of the supported providers, see LabKit, but as an example, for Jaeger tracing support, include the tags: `BUILD_TAGS="tracer_static tracer_static_jaeger"`.

```shell
make BUILD_TAGS="tracer_static tracer_static_jaeger"
```

Once Workhorse is compiled with an opentracing provider, the tracing configuration is configured via the `GITLAB_TRACING` environment variable.

For example:

```shell
GITLAB_TRACING=opentracing://jaeger ./gitlab-workhorse
```

## License

This code is distributed under the MIT license, see the LICENSE file.

[brief-history-blog]: https://about.gitlab.com/2016/04/12/a-brief-history-of-gitlab-workhorse/


CHANGELOG
# Changelog for gitlab-workhorse

## v8.42.1

### Fixed
- Fix nil pointer exception when no object storage config is defined
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/565

## v8.42.0

### Added
- Resize images on-demand with `gm convert`
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/546

## v8.41.0

### Added
- Add Azure blob store support
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/555

## v8.40.0

### Added
- Add project level route for conan package uploads
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/558

### Other
- Refactor uploaders to use different upload strategies
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/553

## v8.39.0

### Fixed
- Fix HTTP Range Requests not working on some S3 providers
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/549

### Other
- Vendor httprs module
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/550

### Performance
- Cache references in file
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/544

## v8.38.0

### Added
- Added configuration option PropagateCorrelationID
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/529
  Contributed by Mahmoud Rahbar Azad
- Add support for AWS S3 Server Side Encryption (SSE-KMS)
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/537

### Changed
- Drop Go v1.12 support
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/532

## v8.37.0

- No changes.
## v8.36.0

- No changes.
## v8.35.0

### Fixed
- Fix Content-Length set prior to SendUrl injection
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/496
  Contributed by Georges-Etienne Legendre

## v8.34.0

### Added
- Support Workhorse directly uploading files to S3
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/466

### Fixed
- Disable compression for open archive
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/508
  Contributed by Georges-Etienne Legendre

### Other
- Add configuration to support an S3 client inside Workhorse
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/516
- Refactor Preparer and SaveFileOpts handling
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/515

## v8.33.0

### Added
- Add routes for Group import via the UI
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/493

### Fixed
- Gather gitlab-zip-cat/metadata stderr and log output in the current context
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/498
  Contributed by Georges-Etienne Legendre

## v8.32.1

### Security
- Limit memory footprint of a command that generates ZIP artifacts metadata
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/

## v8.32.0

### Added
- Process LSIF document before sending it to GitLab
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/492
- Delay PostUploadPack response until request is fully read
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/494  

## v8.31.1

### Security
- Limit memory footprint of a command that generates ZIP artifacts metadata
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/

## v8.31.0

### Added
- Add a signed field on upload requests containing all the workhorse parameters
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/490

### Other
- Add automatic changelog generation
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/484

## v8.30.2

### Security
- Limit memory footprint of a command that generates ZIP artifacts metadata
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/

v 8.30.1

- Sign artifact multipart fields in Workhorse

v 8.30.0

- Proxy ActionCable websocket connection !454

v 8.29.0

- Bump Labkit version to support Profiler sample versioning !479

v 8.28.0

- Reject parameters that override upload fields
- PyPi - Object storage upload route for package files !474

v 8.27.0

- Remove Set-Cookie header from archive and raw blob responses !475

v 8.26.0

- Add route for project imports direct upload via UI !470

## v8.25.3

### Security
- Limit memory footprint of a command that generates ZIP artifacts metadata
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/

### Other
- Add automatic changelog generation
  https://gitlab.com/gitlab-org/gitlab-workhorse/-/merge_requests/484

v 8.25.2

- Sign artifact multipart fields in Workhorse

v 8.25.1

- Reject parameters that override upload fields

v 8.25.0

- Add route for project imports direct upload !459

v 8.24.0

- Support Stackdriver Profiler through Labkit !461

v 8.23.0

- Don't set Cache-Control header for git archives !462

v 8.22.0

- Bump the version of golang.org/x/sys !456
- Add friendly development error page for 502 !453

v 8.21.2

- Sign artifact multipart fields in Workhorse

v 8.21.1

- Reject parameters that override upload fields

v 8.21.0

- Add route for group imports direct upload !455

v 8.20.2

- Sign artifact multipart fields in Workhorse

v 8.20.1

- Reject parameters that override upload fields

v 8.20.0

- Sign file upload requests modified by workhorse

v 8.19.0

- Use multipart uploads for nuget packages !451

v 8.18.0

- Allow inline content disposition for pdf files !446
- Update environment websocket route !449

v 8.17.0

- Add labkit monitoring for build metrics !440
- Log duration_ms when Workhorse hits a bad gateway !445

v 8.16.0

- Ignore CompleteMultipartUpload ETag !438
- Add NuGet route for package uploads !441
- Upgrade Gitaly client to v1.74.0 !443
- Set a time limit on git upload-pack requests

v 8.15.0

- Object store case insensitive ETag comparison !434
- Upgrade gitaly to 1.68.0 !435

v 8.14.1

- Set a time limit on git upload-pack requests

v 8.14.0

- Keep HTTP 1.0 cache headers from sendurl proxies !431

v 8.13.0


main.go
/*
gitlab-workhorse handles slow requests for GitLab

This HTTP server can service 'git clone', 'git push' etc. commands
from Git clients that use the 'smart' Git HTTP protocol (git-upload-pack
and git-receive-pack). It is intended to be deployed behind NGINX
(for request routing and SSL termination) with access to a GitLab
backend (for authentication and authorization) and local disk access
to Git repositories managed by GitLab. In GitLab, this role was previously
performed by gitlab-grack.

In this file we start the web server and hand off to the upstream type.
*/
package main

import (
	"context"
	"flag"
	"fmt"
	"net"
	"net/http"
	_ "net/http/pprof"
	"os"
	"syscall"
	"time"

	"gitlab.com/gitlab-org/labkit/log"
	"gitlab.com/gitlab-org/labkit/monitoring"
	"gitlab.com/gitlab-org/labkit/tracing"

	"gitlab.com/gitlab-org/gitlab-workhorse/internal/config"
	"gitlab.com/gitlab-org/gitlab-workhorse/internal/queueing"
	"gitlab.com/gitlab-org/gitlab-workhorse/internal/redis"
	"gitlab.com/gitlab-org/gitlab-workhorse/internal/secret"
	"gitlab.com/gitlab-org/gitlab-workhorse/internal/upstream"
)

// Version is the current version of GitLab Workhorse
var Version = "(unknown version)" // Set at build time in the Makefile
// BuildTime signifies the time the binary was build
var BuildTime = "19700101.000000" // Set at build time in the Makefile

var printVersion = flag.Bool("version", false, "Print version and exit")
var configFile = flag.String("config", "", "TOML file to load config from")
var listenAddr = flag.String("listenAddr", "localhost:8181", "Listen address for HTTP server")
var listenNetwork = flag.String("listenNetwork", "tcp", "Listen 'network' (tcp, tcp4, tcp6, unix)")
var listenUmask = flag.Int("listenUmask", 0, "Umask for Unix socket")
var authBackend = flag.String("authBackend", upstream.DefaultBackend.String(), "Authentication/authorization backend")
var authSocket = flag.String("authSocket", "", "Optional: Unix domain socket to dial authBackend at")
var cableBackend = flag.String("cableBackend", upstream.DefaultBackend.String(), "ActionCable backend")
var cableSocket = flag.String("cableSocket", "", "Optional: Unix domain socket to dial cableBackend at")
var pprofListenAddr = flag.String("pprofListenAddr", "", "pprof listening address, e.g. 'localhost:6060'")
var documentRoot = flag.String("documentRoot", "public", "Path to static files content")
var proxyHeadersTimeout = flag.Duration("proxyHeadersTimeout", 5*time.Minute, "How long to wait for response headers when proxying the request")
var developmentMode = flag.Bool("developmentMode", false, "Allow the assets to be served from Rails app")
var secretPath = flag.String("secretPath", "./.gitlab_workhorse_secret", "File with secret key to authenticate with authBackend")
var apiLimit = flag.Uint("apiLimit", 0, "Number of API requests allowed at single time")
var apiQueueLimit = flag.Uint("apiQueueLimit", 0, "Number of API requests allowed to be queued")
var apiQueueTimeout = flag.Duration("apiQueueDuration", queueing.DefaultTimeout, "Maximum queueing duration of requests")
var apiCiLongPollingDuration = flag.Duration("apiCiLongPollingDuration", 50, "Long polling duration for job requesting for runners (default 50s - enabled)")

var prometheusListenAddr = flag.String("prometheusListenAddr", "", "Prometheus listening address, e.g. 'localhost:9229'")

var logConfig = logConfiguration{}

func init() {
	flag.StringVar(&logConfig.logFile, "logFile", "", "Log file location")
	flag.StringVar(&logConfig.logFormat, "logFormat", "text", "Log format to use defaults to text (text, json, structured, none)")
}

func main() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage of %s:\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "\n  %s [OPTIONS]\n\nOptions:\n", os.Args[0])
		flag.PrintDefaults()
	}
	kgb := &kgbapp.App{}
	kgbapp.BindFlags(flag.CommandLine, kgb, "kgb-")
	flag.Parse()

	if *printVersion {
		fmt.Printf("gitlab-workhorse %s-%s\n", Version, BuildTime)
		os.Exit(0)
	}

	closer, err := startLogging(logConfig)
	if err != nil {
		log.WithError(err).Fatal("Unable to configure logger")
	}
	defer closer.Close()

	tracing.Initialize(tracing.WithServiceName("gitlab-workhorse"))

	backendURL, err := parseAuthBackend(*authBackend)
	if err != nil {
		log.WithError(err).Fatal("Invalid authBackend")
	}

	cableBackendURL, err := parseAuthBackend(*cableBackend)
	if err != nil {
		log.WithError(err).Fatal("Invalid cableBackend")
	}

	log.WithField("version", Version).WithField("build_time", BuildTime).Print("Starting")

	// Good housekeeping for Unix sockets: unlink before binding
	if *listenNetwork == "unix" {
		if err := os.Remove(*listenAddr); err != nil && !os.IsNotExist(err) {
			log.WithError(err).Fatal("Failed to remove socket")
		}
	}

	// Change the umask only around net.Listen()
	oldUmask := syscall.Umask(*listenUmask)
	listener, err := net.Listen(*listenNetwork, *listenAddr)
	syscall.Umask(oldUmask)
	if err != nil {
		log.WithError(err).Fatal("Failed to listen")
	}

	// The profiler will only be activated by HTTP requests. HTTP
	// requests can only reach the profiler if we start a listener. So by
	// having no profiler HTTP listener by default, the profiler is
	// effectively disabled by default.
	if *pprofListenAddr != "" {
		go func() {
			err := http.ListenAndServe(*pprofListenAddr, nil)
			if err != nil {
				log.WithError(err).Error("Failed to start pprof listener")
			}
		}()
	}

	monitoringOpts := []monitoring.Option{monitoring.WithBuildInformation(Version, BuildTime)}

	if *prometheusListenAddr != "" {
		monitoringOpts = append(monitoringOpts, monitoring.WithListenerAddress(*prometheusListenAddr))
	}

	go func() {
		err := monitoring.Start(monitoringOpts...)
		if err != nil {
			log.WithError(err).Error("Failed to start monitoring")
		}
	}()

	secret.SetPath(*secretPath)
	cfg := config.Config{
		Backend:                  backendURL,
		CableBackend:             cableBackendURL,
		Socket:                   *authSocket,
		CableSocket:              *cableSocket,
		Version:                  Version,
		DocumentRoot:             *documentRoot,
		DevelopmentMode:          *developmentMode,
		ProxyHeadersTimeout:      *proxyHeadersTimeout,
		APILimit:                 *apiLimit,
		APIQueueLimit:            *apiQueueLimit,
		APIQueueTimeout:          *apiQueueTimeout,
		APICILongPollingDuration: *apiCiLongPollingDuration,
	}

	if *configFile != "" {
		cfgFromFile, err := config.LoadConfig(*configFile)
		if err != nil {
			log.WithField("configFile", *configFile).WithError(err).Fatal("Can not load config file")
		}

		cfg.Redis = cfgFromFile.Redis
		cfg.ObjectStorageCredentials = cfgFromFile.ObjectStorageCredentials

		if cfg.Redis != nil {
			redis.Configure(cfg.Redis, redis.DefaultDialFunc)
			go redis.Process()
		}
	}

	accessLogger, accessCloser, err := getAccessLogger(logConfig)
	if err != nil {
		log.WithError(err).Fatal("Unable to configure access logger")
	}
	if accessCloser != nil {
		defer accessCloser.Close()
	}

	up := wrapRaven(upstream.NewUpstream(cfg, accessLogger))

	go func() {
		err := kgb.Run(context.TODO())
		if err != nil {
			log.WithError(err).Fatal("Failed to start kgb")
		}
	}()

	err = http.Serve(listener, up)
	if err != nil {
		log.WithError(err).Fatal("Unable to serve")
	}
}
